<a href="https://colab.research.google.com/github/gcochlar/Bootcamp_DataScience/blob/main/Modulo_04/Bootcamp_Mod4_Aula05.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://drive.google.com/uc?id=16O9eMrtSeRDnDpZgmneXj34eIClCntvg'>

# **Tratamento, Análise e** ***Machine Learning*** **Aplicado**

---

Esse notebook servirá para o acompanhamento das aulas do Módulo 04 do ***Bootcamp de Data Science Aplicada*** realizado pela Alura.

Este módulo dará início ao desenvolvimento de um projeto prático que passará por todo o *workflow* em *Data Science*, do entendimento do problema, tratamento e análise dos dados até a proposta de solução utilizando ***Machine Learning***.

Vamos trabalhar com dados da **COVID-19** do **Hospital Sírio Libanês**, focando este módulo no tratamento e análise de dados para entender profundamente o problema que estamos lidando e propor possíveis soluções.

---

## **Aula 01 -** ***Machine Learning*** **e Saúde**
---


In [ ]:
import pandas as pd


In [ ]:
url = 'https://github.com/gcochlar/Bootcamp_DataScience/blob/main/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true'

In [ ]:
dados = pd.read_excel(url)

In [ ]:
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

A primeira aula serviu para explicar o funcionamento do **Kaggle** e nos ajudar a localizar o desafio do **Hospital Sírio Libanês** que está postado lá ([link aqui](https://www.kaggle.com/S%C3%ADrio-Libanes/covid19)).

Também nos mostrou como interpretar inicialmente os dados do desafio e como criar e trabalhar com o notebook diretamente dentro do **Kaggle**.

Além disso, foi explicada a diferença entre ***Algoritmo de Classificação Binária***, que será utilizado nesse desafio e ***Algoritmo de Regressão***.

Na ***Classificação Binária*** buscamos uma resposta *booleana*, o resultado acontece ou não (1 e 0). No presente caso, o paciente vai precisar de internação em UTI ou não.

Na ***Regressão*** buscamos um resultado numa faixa contínua de resultados. Por exemplo, um preço de venda para um apartamento.

Em função da continuidade e para continuar postando os notebooks no **GitHub**, vou continuar usando o **Colab**.

#### **Desafio - Explorar o Kaggle, os datasets, competições e discussões para verificar o quão rico é esta plataforma**
---

Já conhecia o **Kaggle** desde a *Imersão em Data Science* da **Alura**, mas aproveitei a oportunidade para navegar mais um pouco pelo site.

Achei interessante a disponibilidade de bases de dados de assuntos diversos como cinema, quadrinhos e música, além de assuntos mais sérios como já era esperado.

Não tinha notado também que o **Kaggle** oferece alguns cursos, pretendo retornar posteriormente para ver o que eles disponibilizam em termos de ***Machine Learning***.

#### **Desafio - Leia a descrição do problema proposto pelo Sírio Libanês, tentando entender qual é o real problema. Antes de continuar para as próximas aulas, debata no Discord suas conclusões sobre o projeto**
---

O problema proposto no desafio estabelece duas tarefas diferentes. Vi que no **Discord** houve uma certa discussão a respeito de serem a mesma coisa ou não, mas o meu entendimento é de que são tarefas diferentes mesmo.

>**Task 01**
>
>Predict admission to the ICU of confirmed COVID-19 cases.
>
>Based on the data available, is it feasible to predict which patients will need intensive care unit support?
>
>The aim is to provide tertiary and quarternary hospitals with the most accurate answer, so ICU resources can be arranged or patient transfer can be scheduled.

Essa tarefa busca maximizar o acerto de casos positivos minimizando os falsos negativos, ou seja, acertar o máximo de casos de pacientes que irão precisar de UTI, para evitar o sub- ou superdimensionamento da estrutura.

>**Task 02**
>
>Predict NOT admission to the ICU of confirmed COVID-19 cases.
>
>Based on the subsample of widely available data, is it feasible to predict which patients will need intensive care unit support?
>
>The aim is to provide local and temporary hospitals a good enough answer, so frontline physicians can safely discharge and remotely follow up with these patients.

Essa tarefa busca maximizar o acerto de casos negativos minimizando os falsos positivos, ou seja, acertar o máximo de casos que não irão precisar de UTI e que poderiam receber alta e ser monitorados à distância, liberando a estrutura hospitalar para receber mais pacientes.

### **O que aprendemos nessa aula?**
---

>* O que é o **Kaggle** e como utilizar;
* Como interpretar o problema proposto pelo **Sírio Libanês**;
* O que é uma Classificação Binária.

---


## **Aula 02 - Desenvolvimento no** ***Kaggle***
---


Apesar da aula continuar usando o **Kaggle**, vou permanecer no **Colab**.

Vamos começar analisando o que temos na base de dados importada.

In [ ]:
dados.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

In [ ]:
dados['ICU'].value_counts()

0    1410
1     515
Name: ICU, dtype: int64

#### **Desafio - Pesquisar Aprendizado Supervisionado, Não Supervisionado, Por Reforço e Mais**
---

**Aprendizado Supervisionado**

É o modelo mais comum. Treinamos o algoritmo com um volume inicial de dados, mostrando características e o resultado obtido. O processo de aprendizado leva a prever a saída a partir de novos dados de entrada.

**Aprendizado Não Supervisionada**

É o modelo utilizado quando não temos os dados de resultado obtido na base de dados inicial. O algoritmo tenta entender as características e gerar agrupamentos (*clusters*), muitas vezes para depois passar cada *cluster* por um processo de aprendizado supervisionado.

**Aprendizado Por Reforço**

É o modelo mais indicado para tomada de decisões. Funciona por tentativa e erro, onde a máquina é exposta a vários processos decisórios e vai aprendendo conforme as decisões tomadas forem acertadas ou não.

**Redes Neurais e** ***Deep Learning***

Trabalha com sucessivas camadas de aprendizado para melhorar o processo de tomada de decisão. É utilizado para descobrir padrões em dados desestruturados, ideal para reconhecimento de imagens e de fala, por exemplo.

Normalmente as camadas são formadas por combinações de algoritmos de aprendizado supervisionado e não supervisionado.

Fontes:

[Data Hackers](https://medium.com/data-hackers/simplificando-as-categorias-de-algoritmos-de-machine-learning-191ac9dc8ddf)

[IBM - Machine Learning for Dummies](https://www.ibm.com/downloads/cas/GB8ZMQZ3)

[LAMFO - UNB](https://lamfo-unb.github.io/2017/07/27/tres-tipos-am/)


#### **Desafio - Pesquisar Sobre Regressão, Classificação, etc.**
---

A diferença entre a **Regressão** e a **Classificação** se dá normalmente pelo tipo de resultado que teremos.

Para modelos de **Classificação** teremos um número limitado e discreto de resultados.

Já para modelos de **Regressão** o resultado esperado está numa faixa contínua de dados. É mais utilizado para encontrar a correlação entre variáveis.

### **O que aprendemos nessa aula?**
---

>* Como usar os notebooks direto do **Kaggle**;
>
>* O que são algoritmos supervisionados;
>
>* O que são dados anonimizados e sua relevância da perspectiva ética.

---


## **Aula 03 - Modelos de** ***Machine Learning***
---


Para iniciar o nosso estudo, vamos usar a biblioteca **```SciKitLearn```** e o seu modelo de Regressão Logística.

In [ ]:
from sklearn.linear_model import LogisticRegression

modelo = LogisticRegression()

modelo

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Vamos preparar a nossa base de dados para manter só as colunas com números, deixar de fora as linhas com dados nulos e NaN e também a coluna 'ICU', que é a coluna que contém o resultado.

Essa coluna 'ICU' vai para uma tabela em separado, para avaliarmos o resultado depois.

In [ ]:
# Pegar só as colunas que contém números
x_columns = dados.describe().columns

# Excluir linhas que contém NaN
dados_limpos = dados.dropna()

# Separar os dados de resultado (y) e de entrada (x)
y = dados_limpos.ICU
x = dados_limpos[x_columns].drop(['ICU'], axis = 1)

In [ ]:
modelo.fit(x,y)
prev = modelo.predict(x)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [ ]:
print(sum(prev == y),'acertos')
print(sum(prev != y),'erros')
print(sum(prev == y) / len(x) * 100,'% de acertos')
print(sum(prev != y) / len(x) * 100,'% de erros')

497 acertos
48 erros
91.19266055045871 % de acertos
8.807339449541285 % de erros


#### **Desafio - Pesquisar Sobre Ética em** ***Data Science*** **e Inteligência Artificial**
---



https://insider.dn.pt/em-rede/privacidade-no-data-science-a-etica-deve-ir-muito-alem-da-lei/25736/

https://www.serpro.gov.br/lgpd/noticias/2019/tratamento-etico-dados-pessoais-ciencia-data-science

http://www.sciencefriday.com/articles/10-questions-for-the-nations-first-chief-data-scientist/

https://medium.com/somos-tera/o-que-e-etica-de-dados-3ce4702cdfc1

https://royalsocietypublishing.org/doi/full/10.1098/rsta.2016.0360#sec-1

https://www.itinsight.pt/news/data-and-analytics/insights-2018---etica-estrategia-e-futuro-da-data-science-em-debate

https://neofeed.com.br/blog/home/qual-e-a-etica-da-inteligencia-artificial/

https://pt.wikipedia.org/wiki/%C3%89tica_na_intelig%C3%AAncia_artificial

https://www.umov.me/inteligencia-artificial-etica-e-trabalho/



#### **Desafio - Qual o motivo de jogar fora o ICU dos dados X? O que acontece se estiver nos dados X? Isso seria bom ou ruim?**
---



Como podemos ver no resultado do código abaixo, se rodarmos o modelo com os dados da coluna 'ICU' inclusas no parâmetro X, o modelo será treinado para responder com 100% de acerto para esses dados de treino, pois ele acaba aprendendo que sempre que receber no *input* que 'ICU' = 1, ele vai responder que o *output*, que também é 'ICU' será 1, o que é óbvio.

Mas esse modelo não vai ter nenhuma utilidade, pois vai depender de termos a resposta da coluna 'ICU' para nos dar a previsão de resposta (que já teremos). Passa a ser um modelo de previsão do passado.

In [ ]:
modelo2 = LogisticRegression();
x2 = dados_limpos[x_columns]
modelo2.fit(x2,y);
print(sum(modelo2.predict(x2) == y),'acertos');
print(sum(modelo2.predict(x2) != y),'erros');
print(sum(modelo2.predict(x2) == y) / len(x2) * 100,'% de acertos');
print(sum(modelo2.predict(x2) != y) / len(x2) * 100,'% de erros');

545 acertos
0 erros
100.0 % de acertos
0.0 % de erros


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### **O que aprendemos nessa aula?**
---

>* O que é a uma regressão logística;
>
>* Como treinar o seu primeiro modelo de ***Machine Learning***.

---


## **Aula 04 - Métricas e Avaliações**
---


Vamos melhorar o nosso modelo, mas antes vamos ver qual seria a classificação de um modelo **```DummyClassifier```**, que simplesmente classifica tudo com base na proporção que temos na base de dados.

Vamos verificar qual a proporção da base de dados.

In [ ]:
y.value_counts()

0    281
1    264
Name: ICU, dtype: int64

Vemos que o resultado 0 está em 281 linhas, ou seja, 51,56% da base.

In [ ]:
from sklearn.dummy import DummyClassifier

modelo_dummy = DummyClassifier()
modelo_dummy.fit(x,y)
prev_dummy = modelo_dummy.predict(x)
print(sum(prev_dummy == y) / len(x) * 100,'% de acertos');

48.440366972477065 % de acertos


/usr/local/lib/python3.7/dist-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


In [ ]:
import numpy as np

teste = np.bincount(prev_dummy) ## Equivalente ao value_counts para arrays
ii = np.nonzero(teste)[0]

np.vstack((ii,teste[ii])).T

array([[  0, 276],
       [  1, 269]])

A biblioteca **```sklearn```** também tem uma ferramenta de avaliação de acurácia do modelo, para nos permitir comparações.

No presente caso, a resposta vai ser a mesma que já calculamos acima.

In [ ]:
from sklearn.metrics import accuracy_score

accuracy_score(y, prev_dummy)

0.48440366972477067

In [ ]:
accuracy_score(y, modelo2.predict(x2))

1.0

In [ ]:
accuracy_score(y, prev)

0.9119266055045872

O problema de fazermos o treino com toda a base é que isso aumenta artificialmente a acurácia do modelo, pois ele vai ser testado contra dados que também foram usados no treino.

Para que o modelo esteja preparado para trabalhar com dados novos, precisamos testar a acurácia contra dados que não foram usados no treino.

Para isso, vamos agora fazer a separação dos nossos dados em treino e teste, para poder criar um modelo e depois verificar a acurácia contra dados que ele nunca viu antes.

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y)

In [ ]:
print(len(x_train), len(y_train))
print(len(x_test), len(y_test))

408 408
137 137


In [ ]:
modelo_2 = LogisticRegression()
modelo_2.fit(x_train, y_train)

y_prev = modelo_2.predict(x_test)
accuracy_score(y_test, y_prev)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8978102189781022

In [ ]:
np.random.seed(73246)

x3_train, x3_test, y3_train, y3_test = train_test_split(x, y, stratify = y)
print(len(x3_train), len(y3_train))
print(len(x3_test), len(y3_test))

408 408
137 137


In [ ]:
modelo_3 = LogisticRegression()
modelo_3.fit(x3_train, y3_train)

y3_prev = modelo_3.predict(x3_test)
accuracy_score(y3_test, y3_prev)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8248175182481752

#### **Desafio - Decida quanto você vai escolher para treino e quanto para teste? Quanto as pessoas utilizam e qual o motivo?**
---

A divisão mais comum é na proporção 70%\~30% entre dados de treino e de teste. Gosto de usar essa proporção sempre que possível. No caso específico do COVID-19 tenho encontrado dificuldade para utilizar essa proporção em razão da pequena quantidade de dados disponíveis.

No módulo anterior foi utilizado quase 84% (400 das 478 linhas) dos dados para treino, justamente por causa disso.

A função **```train_test_split```** faz a divisão tendo como parâmetro *default* a proporção de 75%\~25%, mas isso pode ser alterado mexendo no parâmetro **```test_size```**, que recebe um número entre 0 e 1 que vai indicar a proporção do teste (o resto é treino).

Para fazer uma divisão entre dados de treino, de teste e de validação, encontrei a função abaixo.


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

def split_stratified_into_train_val_test(df_input, stratify_colname='y',
                                         frac_train=0.6, frac_val=0.15, frac_test=0.25,
                                         random_state=None):
    '''
    Splits a Pandas dataframe into three subsets (train, val, and test)
    following fractional ratios provided by the user, where each subset is
    stratified by the values in a specific column (that is, each subset has
    the same relative frequency of the values in the column). It performs this
    splitting by running train_test_split() twice.

    Parameters
    ----------
    df_input : Pandas dataframe
        Input dataframe to be split.
    stratify_colname : str
        The name of the column that will be used for stratification. Usually
        this column would be for the label.
    frac_train : float
    frac_val   : float
    frac_test  : float
        The ratios with which the dataframe will be split into train, val, and
        test data. The values should be expressed as float fractions and should
        sum to 1.0.
    random_state : int, None, or RandomStateInstance
        Value to be passed to train_test_split().

    Returns
    -------
    df_train, df_val, df_test :
        Dataframes containing the three splits.
    '''

    if frac_train + frac_val + frac_test != 1.0:
        raise ValueError('fractions %f, %f, %f do not add up to 1.0' % \
                         (frac_train, frac_val, frac_test))

    if stratify_colname not in df_input.columns:
        raise ValueError('%s is not a column in the dataframe' % (stratify_colname))

    X = df_input # Contains all columns.
    y = df_input[[stratify_colname]] # Dataframe of just the column on which to stratify.

    # Split original dataframe into train and temp dataframes.
    df_train, df_temp, y_train, y_temp = train_test_split(X,
                                                          y,
                                                          stratify=y,
                                                          test_size=(1.0 - frac_train),
                                                          random_state=random_state)

    # Split the temp dataframe into val and test dataframes.
    relative_frac_test = frac_test / (frac_val + frac_test)
    df_val, df_test, y_val, y_test = train_test_split(df_temp,
                                                      y_temp,
                                                      stratify=y_temp,
                                                      test_size=relative_frac_test,
                                                      random_state=random_state)

    assert len(df_input) == len(df_train) + len(df_val) + len(df_test)

    return df_train, df_val, df_test

#### **Desafio - Olhe a documentação da ```LogisticRegression```... o que tem de parâmetro em comum com o ```train_test_split``` e o que isso significa para nós?**
---

O parâmetro que ambos têm em comum é o **```random_state```**.

Acredito que sirva para garantir a reprodutibilidade do estudo, mas no caso da Regressão Logística, esse parâmetro só tem efeito para alguns algoritmos (e o *default* não está entre eles).

#### **OBS.:** Não consegui atingir o mesmo resultado que o vídeo da aula.

Copiei o código da aula disponibilizado abaixo, idêntico, em uma célula única, abaixo. Mesmo assim, não fecha o resultado, não entendi...

In [ ]:
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

import pandas as pd

dados = pd.read_excel("https://github.com/gcochlar/Bootcamp_DataScience/blob/main/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true")
dados.head()

dados.describe()

dados["ICU"]

dados["ICU"].value_counts()

x_columns = dados.describe().columns

dados_limpos = dados.dropna()

y = dados_limpos["ICU"]
x = dados_limpos[x_columns].drop(["ICU"], axis=1)

from sklearn.linear_model import LogisticRegression

modelo = LogisticRegression()
modelo.fit(x, y)

modelo.predict([x.iloc[4]])

x.iloc[4]

y.iloc[4]

sum(modelo.predict(x) == y)

sum(modelo.predict(x) != y)

accuracy = sum(modelo.predict(x) == y) / len(y) * 100
accuracy

y.value_counts()

from sklearn.dummy import DummyClassifier

modelo = DummyClassifier("most_frequent")
modelo.fit(x, y)
sum(modelo.predict(x) == y) / len(y) * 100

from sklearn.metrics import accuracy_score

accuracy_score(y, modelo.predict(x))

import numpy as np

np.random.seed(73246)

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

print(len(x_train), len(y_train))
print(len(x_test), len(y_test))

modelo = LogisticRegression()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)


408 408
137 137


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


0.8248175182481752

No *notebook* que tem só o conteúdo da aula, o resultado é 0.8248175182481752.

A mesma célula, copiada para cá, dá esse resultado acima (0.8321167883211679). Isso não tem explicação...

E um outro teste: reiniciei o ambiente de execução e executei somente essa célula acima. Daí consegui obter o resultado da aula.

Tem alguma coisa no código que afeta o resultado, não consegui localizar o que...

### **O que aprendemos nessa aula?**
---

>* Como metrificar um modelos de classificação, usando o **```Sci-kit Learn```**;
>
>* Como realizar o treinamento dividindo em dados de treino e teste.

---


## **Aula 05 - Compreendendo o Problema Real**
---


Retomando desde o início... vamos voltar ao início da exploração dos dados, agora analisando alguns pontos que, por motivos didáticos, ficaram de lado no início da análise.

In [ ]:
dados = pd.read_excel(url)
dados.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

In [ ]:
dados.columns

Index(['PATIENT_VISIT_IDENTIFIER', 'AGE_ABOVE65', 'AGE_PERCENTIL', 'GENDER',
       'DISEASE GROUPING 1', 'DISEASE GROUPING 2', 'DISEASE GROUPING 3',
       'DISEASE GROUPING 4', 'DISEASE GROUPING 5', 'DISEASE GROUPING 6',
       ...
       'TEMPERATURE_DIFF', 'OXYGEN_SATURATION_DIFF',
       'BLOODPRESSURE_DIASTOLIC_DIFF_REL', 'BLOODPRESSURE_SISTOLIC_DIFF_REL',
       'HEART_RATE_DIFF_REL', 'RESPIRATORY_RATE_DIFF_REL',
       'TEMPERATURE_DIFF_REL', 'OXYGEN_SATURATION_DIFF_REL', 'WINDOW', 'ICU'],
      dtype='object', length=231)

In [ ]:
dados.WINDOW.unique()

array(['0-2', '2-4', '4-6', '6-12', 'ABOVE_12'], dtype=object)

O ponto mais importante que ficou de fora na primeira rodada de exploração dos dados foi o fato de que temos 1925 linhas na tabela que são 385 pacientes (ver mínimo e máximo da coluna **```PATIENT_VISIT_IDENTIFIER```**) com 5 linhas para cada um em função das 5 janelas de tempo de medições (coluna **```WINDOW```**, com intervalos de 0-2h, 2-4h, 4-6h, 6-12h e acima de 12h).

Sabendo isso, precisamos tomar algumas providências adicionais de limpeza em nossa base de dados.

In [ ]:
dados.iloc[:,-2:].head(20)

WINDOW  ICU
0        0-2    0
1        2-4    0
2        4-6    0
3       6-12    0
4   ABOVE_12    1
5        0-2    1
6        2-4    1
7        4-6    1
8       6-12    1
9   ABOVE_12    1
10       0-2    0
11       2-4    0
12       4-6    0
13      6-12    0
14  ABOVE_12    1
15       0-2    0
16       2-4    0
17       4-6    0
18      6-12    0
19  ABOVE_12    0

**Observações / Providências:**
* temos que descartar as linhas (janelas) onde aparece o 1 na ICU, pois as medições podem ter sido feitas depois do paciente já ter sido transferido;
* poderíamos pegar os dados com ICU=0 e verificar se o paciente vai para a UTI no próximo intervalo. Gui disse que tentou esse e não teve resultados interessantes;
* depois que o paciente vai para UTI, continua replicando o 1 na coluna ICU;
* possível caminho seria analisar as colunas de 0-2h e checar a UTI seguinte. Depois, incluir colunas de 2-4h e, com tudo isso (0-4h, o dobro de colunas), checar UTI seguinte e assim por diante. Mas isso traria muitas colunas e teria muita informação em branco nos períodos iniciais, muito lixo na base;
* para fazer previsão, quando tem lacunas na base, um dummy regressor prevê o dado anterior ou posterior para preencher a lacuna, pois é o mais razoável (falando de séries temporais);

Vamos fazer algumas simulações com dados falsos, para entender as lógicas de preenchimento.

#### **Dados** ***Fake***
---

Vamos criar um pequeno *DataFrame* com dados exemplificativos para entender o funcionamento dos diferentes métodos de preenchimento do comando **```fillna```**, que utilizaremos para preencher as lacunas em nosso *dataset*.

In [ ]:
dados_fake = pd.DataFrame([[1,1],[1,None],[2, 3.0],[2,None]], columns=['Paciente', 'Leitura'])
dados_fake

Paciente  Leitura
0         1      1.0
1         1      NaN
2         2      3.0
3         2      NaN

In [ ]:
dados_fake.fillna(method='ffill')

Paciente  Leitura
0         1      1.0
1         1      1.0
2         2      3.0
3         2      3.0

O método **```ffill```** preenche para frente, colocando os dados encontrados para a próxima lacuna.

E se tivermos um paciente sem leitura nenhuma?

In [ ]:
dados_fake = pd.DataFrame([[1,1],[1,None],[2, None],[2,None]], columns=['Paciente', 'Leitura'])
dados_fake

Paciente  Leitura
0         1      1.0
1         1      NaN
2         2      NaN
3         2      NaN

In [ ]:
dados_fake.fillna(method='ffill')

Paciente  Leitura
0         1      1.0
1         1      1.0
2         2      1.0
3         2      1.0

Usando o comando para o *DataFrame* inteiro, sem indicar algum tipo de quebra, o resultado foi que os dados do paciente 1 foram utilizados para o paciente 2, o que não pode acontecer.

Temos que agrupar por paciente, para impedir isso.

In [ ]:
dados_fake = pd.DataFrame([[1,1,1],[1,None,2],[2, None,1],[2,None,2]], columns=['Paciente', 'Leitura', 'Janela'])
dados_fake

Paciente  Leitura  Janela
0         1      1.0       1
1         1      NaN       2
2         2      NaN       1
3         2      NaN       2

In [ ]:
dados_fake.groupby('Paciente').fillna(method='ffill')

Leitura  Janela
0      1.0       1
1      1.0       2
2      NaN       1
3      NaN       2

In [ ]:
dados_fake.groupby('Paciente', as_index=False).fillna(method='ffill')

Leitura  Janela
0      1.0       1
1      1.0       2
2      NaN       1
3      NaN       2

In [ ]:
dados_fake.groupby('Paciente', as_index=True).fillna(method='ffill')

Leitura  Janela
0      1.0       1
1      1.0       2
2      NaN       1
3      NaN       2

Nesse caso podemos ver que o **Python** preencheu corretamente os dados do paciente 1 e, como não encontrou nenhum dado para o paciente 2, ele permaneceu com os dados nulos.

O Guilherme disse para colocar o **`as_index`** como False, mas não notei nenhuma alteração no resultado... postei dúvida no **Discord**, mas ainda não tive nenhuma resposta.

In [ ]:
dados_fake = pd.DataFrame([[1,None,1],[1,2.5,2],
                           [2, 3.0,1],[2,None,2]],
                           columns=['Paciente', 'Leitura', 'Janela'])
dados_fake

Paciente  Leitura  Janela
0         1      NaN       1
1         1      2.5       2
2         2      3.0       1
3         2      NaN       2

In [ ]:
dados_fake.groupby('Paciente', as_index=False).fillna(method='ffill')

Leitura  Janela
0      NaN       1
1      2.5       2
2      3.0       1
3      3.0       2

Nesse exemplo podemos ver que o valor do paciente 2 foi preenchido para frente, mas o paciente 1, que tinha a primeira medição em branco, permaneceu em branco.

Agora precisamos mandar preencher para frente e para trás. Vamos repetir o comando acima, agregando um *back fill* na sequência.

In [ ]:
dados_fake = pd.DataFrame([[1,None,1],[1,2.5,2],
                           [2, 3.0,1],[2,None,2]],
                           columns=['Paciente', 'Leitura', 'Janela'])
dados_fake

Paciente  Leitura  Janela
0         1      NaN       1
1         1      2.5       2
2         2      3.0       1
3         2      NaN       2

In [ ]:
dados_fake.groupby('Paciente', as_index=False).fillna(method='ffill').fillna(method='bfill')

Leitura  Janela
0      2.5       1
1      2.5       2
2      3.0       1
3      3.0       2

Daí tivemos o preenchimento esperado de todas as lacunas em nossa base de dados. Vamos fazer um último teste com uma base maior, com casos diversos.

In [ ]:
dados_fake = pd.DataFrame([[1,None,1],[1,2.5,2],
                           [2, 3.0,1],[2,None,2],
                           [3,None,1],[3, 1,2],[3,None,3],
                           [4,None,1],[4,6,2],[4,None,3],[4,5,4],[4,None,5]],
                           columns=['Paciente', 'Leitura', 'Janela'])
dados_fake

Paciente  Leitura  Janela
0          1      NaN       1
1          1      2.5       2
2          2      3.0       1
3          2      NaN       2
4          3      NaN       1
5          3      1.0       2
6          3      NaN       3
7          4      NaN       1
8          4      6.0       2
9          4      NaN       3
10         4      5.0       4
11         4      NaN       5

In [ ]:
dados_fake.groupby('Paciente', as_index=False).fillna(method='ffill').fillna(method='bfill')

Leitura  Janela
0       2.5       1
1       2.5       2
2       3.0       1
3       3.0       2
4       1.0       1
5       1.0       2
6       1.0       3
7       6.0       1
8       6.0       2
9       6.0       3
10      5.0       4
11      5.0       5

Como podemos ver acima, todos os dados foram preenchidos como esperado, primeiramente fazendo o preenchimento para frente (**```ffill```**) e depois, em cima do resultado obtido, fazendo o preenchimento para trás (**```bfill```**).

### **De volta aos dados do hospital**
---

Vamos retomar os dados do hospital, importando de novo para ter a garantia de que temos a base sem alterações.

In [ ]:
dados = pd.read_excel(url)
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

Vamos agora fazer o preenchimento (*back fill* e *forward fill*) dos dados que estão faltando. Mas isso só serve para os dados contínuos, que já estão separados no *dataset*, da coluna 13 até a antepenúltima (-2), conforme a documentação do desafio.

In [ ]:
dados_continuos = dados.iloc[:,13:-2].columns
dados_continuos

Index(['ALBUMIN_MEDIAN', 'ALBUMIN_MEAN', 'ALBUMIN_MIN', 'ALBUMIN_MAX',
       'ALBUMIN_DIFF', 'BE_ARTERIAL_MEDIAN', 'BE_ARTERIAL_MEAN',
       'BE_ARTERIAL_MIN', 'BE_ARTERIAL_MAX', 'BE_ARTERIAL_DIFF',
       ...
       'HEART_RATE_DIFF', 'RESPIRATORY_RATE_DIFF', 'TEMPERATURE_DIFF',
       'OXYGEN_SATURATION_DIFF', 'BLOODPRESSURE_DIASTOLIC_DIFF_REL',
       'BLOODPRESSURE_SISTOLIC_DIFF_REL', 'HEART_RATE_DIFF_REL',
       'RESPIRATORY_RATE_DIFF_REL', 'TEMPERATURE_DIFF_REL',
       'OXYGEN_SATURATION_DIFF_REL'],
      dtype='object', length=216)

In [ ]:
dados.groupby('PATIENT_VISIT_IDENTIFIER', as_index=False)[dados_continuos].fillna(method='bfill').fillna(method='ffill')

ALBUMIN_MEDIAN  ...  OXYGEN_SATURATION_DIFF_REL
0           0.605263  ...                   -1.000000
1           0.605263  ...                   -1.000000
2           0.605263  ...                   -1.000000
3           0.000000  ...                   -1.000000
4           0.000000  ...                   -0.814433
...              ...  ...                         ...
1920        0.605263  ...                   -1.000000
1921        0.605263  ...                   -1.000000
1922        0.605263  ...                   -1.000000
1923        0.605263  ...                   -1.000000
1924        0.605263  ...                   -0.835052

[1925 rows x 216 columns]

Como podemos ver, foi feito o preenchimento das informações que estavam nas colunas de dados contínuos. Mas esse comando não altera diretamente a base de dados, ele simplesmente nos mostra como ficaria.

Para que o resultado passe a ser a nossa base de dados, precisamos atribuir o resultado para o *DataFrame*.

Para fazer isso e já colocar na base de dados os demais dados categóricos, vamos montar uma função específica para fazer todo esse processo de limpeza da base de dados.

In [ ]:
def preenche_tabela(dados):
    dados_continuos_colunas = dados.iloc[:,13:-2].columns ## as colunas com dados contínuos, entre a 13 e a penúltima
    dados_continuos = dados.groupby('PATIENT_VISIT_IDENTIFIER', as_index=False)[dados_continuos_colunas].fillna(method='bfill').fillna(method='ffill')
    dados_categoricos = dados.iloc[:,:13] ## as primeiras 13 colunas
    saida = dados.iloc[:,-2:] ## as duas últimas colunas (WINDOW e ICU)
    dados_finais = pd.concat([dados_categoricos, dados_continuos, saida], ignore_index = True, axis = 1)
    dados_finais.columns = dados.columns 
    return(dados_finais)

In [ ]:
dados_limpos = preenche_tabela(dados)
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

In [ ]:
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

In [ ]:
teste = dados_limpos.iloc[:,:13]
teste.to_excel('teste.xlsx')

teste['X9'] = dados_limpos.isna().sum(axis=1)

teste.to_excel('teste2.xlsx')

teste3 = teste.isna()
teste3.to_excel('teste3.xlsx')

Podemos ver que nas colunas **`DISEASE GROUPING X`** temos um paciente que não tem dados preenchidos, por isso precisamos eliminá-lo da tabela.

In [ ]:
dados_limpos = dados_limpos.dropna()

Agora precisamos remover todos os pacientes que entraram e já foram direto para a UTI nas duas primeiras horas, pois não sabemos se os dados lidos foram antes ou depois da movimentação.

In [ ]:
a_remover = dados_limpos.query('WINDOW == "0-2" and ICU == 1')['PATIENT_VISIT_IDENTIFIER'].values

In [ ]:
dados_limpos = dados_limpos.query('PATIENT_VISIT_IDENTIFIER not in @a_remover')

In [ ]:
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1760.000000  ...  1760.000000
mean                 192.818182  ...     0.201705
std                  110.637724  ...     0.401387
min                    0.000000  ...     0.000000
25%                   97.750000  ...     0.000000
50%                  191.500000  ...     0.000000
75%                  289.250000  ...     0.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

In [ ]:
def prepara_window(rows):
    if(np.any(rows['ICU'] == 1)):
        rows.loc[rows['WINDOW'] == '0-2', 'ICU'] = 1
    return rows.loc[rows['WINDOW'] == '0-2']

dados_limpos = dados_limpos.groupby('PATIENT_VISIT_IDENTIFIER').apply(prepara_window)

In [ ]:
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0
5                        25                         5            0  ...    0-2    0

[5 rows x 231 columns]

In [ ]:
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  ...         ICU
count                352.000000  ...  352.000000
mean                 192.818182  ...    0.463068
std                  110.763735  ...    0.499344
min                    0.000000  ...    0.000000
25%                   97.750000  ...    0.000000
50%                  191.500000  ...    0.000000
75%                  289.250000  ...    1.000000
max                  384.000000  ...    1.000000

[8 rows x 229 columns]

In [ ]:
y = dados_limpos['ICU']
x = dados_limpos[x_columns].drop(['ICU'], axis = 1)

In [ ]:
np.random.seed(73246)

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify = y)

modelo_dummy = DummyClassifier(strategy = 'prior')
modelo_dummy.fit(x_train, y_train)

y_pred_dummy = modelo_dummy.predict(x_test)
accuracy_score(y_test, y_pred_dummy)

0.5340909090909091

In [ ]:
modelo = LogisticRegression(max_iter = 1000)
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.7045454545454546

#### **Desafio - Não Usar** ***back fill*** **da Janela Após a Pessoa Ter Ido para UTI**
---



A questão envolvida nesse desafio é a seguinte:
* Estabelecemos que não queremos usar os dados da janela em que o paciente vai para a UTI, pois as medições podem ter sido feitas após a transferência
* Na sequência em que os comandos de limpeza foram feitos, podemos estar puxando os dados de uma linha em que ICU = 1 para fazer o *back fill* de uma coluna em que o paciente não teve medições mas ainda não estava na UTI
* Melhor alternativa que eu consigo identificar:
 * Partir dos dados completos
 * Criar uma coluna que identifique se o paciente foi para a UTI em algum momento (**`pacte_icu`**)
 * Excluir todas as linhas que tiverem ICU = 1
 * Excluir a coluna ICU original e renomear a nova, para reaproveitamento de funções
 * Executar o *forward fill* e o *back fill*

 Vamos implementar isso abaixo:

Passo 1: Partir dos dados completos

---


In [ ]:
dados_desafio = pd.read_excel(url)

Passo 2: Criar uma coluna que identifique se o paciente foi para a UTI em algum momento (**`pacte_icu`**)

---

In [ ]:
pacte_icu = dados_desafio.groupby('PATIENT_VISIT_IDENTIFIER').max()
dict_pacte_icu = pacte_icu.drop(pacte_icu.iloc[:,0:-1], axis=1).to_dict()['ICU']
dados_desafio['pacte_icu'] = dados_desafio['PATIENT_VISIT_IDENTIFIER'].map(dict_pacte_icu)

Passo 3: Excluir todas as linhas que tiverem ICU = 1

---

In [ ]:
dados_desafio.drop(dados_desafio[dados_desafio.ICU == 1].index, inplace = True)

Passo 4: Excluir a coluna ICU original e renomear a nova, para reaproveitamento de funções

---

In [ ]:
dados_desafio.drop('ICU', axis = 1, inplace = True)
dados_desafio = dados_desafio.rename(columns={'pacte_icu':'ICU'})

Passo 5: Executar o *forward fill* e o *back fill*

---

Como a nova base ficou com a mesma estrutura de colunas do *dataset* original, podemos utilizar a mesma função **`preenche_tabela`** que foi criada anteriormente.

In [ ]:
dados_limpos_desafio = preenche_tabela(dados_desafio)

In [ ]:
dados_limpos_desafio.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1410.000000  ...  1410.000000
mean                 195.193617  ...     0.326241
std                  111.378797  ...     0.469003
min                    0.000000  ...     0.000000
25%                   99.000000  ...     0.000000
50%                  195.000000  ...     0.000000
75%                  294.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

Pela diferença na linha de contagem, podemos ver que continuamos tendo um paciente com os dados inválidos (sem preenchimento) nas colunas **`DISEASE GROUPING X`**.

Retirando esse paciente (usando um **`dropna`**) já podemos rodar novamente a função **`prepara_window`** que vai nos deixar com a base normalizada, com 281 pacientes em linhas únicas, só com os dados da primeira janela (0-2h).

In [ ]:
dados_limpos_desafio = dados_limpos_desafio.dropna()

dados_limpos_desafio = dados_limpos_desafio.groupby('PATIENT_VISIT_IDENTIFIER').apply(prepara_window)

dados_limpos_desafio.describe()

PATIENT_VISIT_IDENTIFIER  ...         ICU
count                352.000000  ...  352.000000
mean                 192.818182  ...    0.463068
std                  110.763735  ...    0.499344
min                    0.000000  ...    0.000000
25%                   97.750000  ...    0.000000
50%                  191.500000  ...    0.000000
75%                  289.250000  ...    1.000000
max                  384.000000  ...    1.000000

[8 rows x 229 columns]

In [ ]:
dados_limpos_desafio.describe()

PATIENT_VISIT_IDENTIFIER  ...         ICU
count                352.000000  ...  352.000000
mean                 192.818182  ...    0.463068
std                  110.763735  ...    0.499344
min                    0.000000  ...    0.000000
25%                   97.750000  ...    0.000000
50%                  191.500000  ...    0.000000
75%                  289.250000  ...    1.000000
max                  384.000000  ...    1.000000

[8 rows x 229 columns]

Ficamos com a mesma quantidade de pacientes válidos.
Agora podemos rodar novamente os modelos e verificar os resultados obtidos.

In [ ]:
y_desafio = dados_limpos_desafio['ICU']
x_desafio = dados_limpos_desafio[x_columns].drop(['ICU'], axis = 1)

In [ ]:
np.random.seed(73246)

In [ ]:
x_desafio_train, x_desafio_test, y_desafio_train, y_desafio_test = train_test_split(x_desafio, y_desafio, stratify = y_desafio)

modelo_desafio_dummy = DummyClassifier(strategy = 'prior')
modelo_desafio_dummy.fit(x_desafio_train, y_desafio_train)

y_desafio_pred_dummy = modelo_desafio_dummy.predict(x_desafio_test)
accuracy_score(y_desafio_test, y_desafio_pred_dummy)

0.5340909090909091

Tivemos exatamente o mesmo resultado para o **`DummyClassifier`**.

In [ ]:
modelo_desafio = LogisticRegression(max_iter = 200)
modelo_desafio.fit(x_desafio_train, y_desafio_train)

y_desafio_prediction = modelo_desafio.predict(x_desafio_test)
accuracy_score(y_desafio_test, y_desafio_prediction)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.6590909090909091

O resultado piorou quase 5 pontos percentuais, caiu de 70,5% para 65,9%.

#### **Desafio - Arranque fora as Colunas que Não Devem Ser Utilizadas**
---



In [ ]:
dados_limpos_desafio['AGE_PERCENTIL'] = dados_limpos_desafio['AGE_PERCENTIL'].str.replace('th', '')
dados_limpos_desafio['AGE_PERCENTIL'] = dados_limpos_desafio['AGE_PERCENTIL'].str.replace('Above ', '')
dados_limpos_desafio['AGE_PERCENTIL'] = dados_limpos_desafio['AGE_PERCENTIL'].astype('int64')

In [ ]:
y_desafio2 = dados_limpos_desafio['ICU']
x_desafio2 = dados_limpos_desafio[x_columns].drop(['ICU'], axis = 1)

np.random.seed(73246)

x_desafio2_train, x_desafio2_test, y_desafio2_train, y_desafio2_test = train_test_split(x_desafio2, y_desafio2, stratify = y_desafio2, test_size=0.3)

modelo_desafio2_dummy = DummyClassifier(strategy = 'prior')
modelo_desafio2_dummy.fit(x_desafio2_train, y_desafio2_train)

y_desafio2_pred_dummy = modelo_desafio2_dummy.predict(x_desafio2_test)
accuracy_score(y_desafio2_test, y_desafio2_pred_dummy)

0.5377358490566038

In [ ]:
modelo_desafio2 = LogisticRegression(max_iter = 10000)
modelo_desafio2.fit(x_desafio2_train, y_desafio2_train)

y_desafio2_prediction = modelo_desafio2.predict(x_desafio2_test)
accuracy_score(y_desafio2_test, y_desafio2_prediction)

0.6509433962264151

Fiz a correção dos dados de idade para numérico, para que o modelo possa usar essa informação, mas não alterou em nada o resultado.

Daí resolvi pegar um grupo qualquer de colunas e tirar fora do meu modelo para fazer um teste. Comecei tirando diversas colunas iniciais (dados categóricos) e vi que não estava afetando em nada o resultado.

Daí resolvi pegar 10 colunas do fim da tabela e deixar de fora também, para ver o efeito nos resultados.

In [ ]:
colunas_a_dropar = ['ICU','GENDER','DISEASE GROUPING 1','DISEASE GROUPING 2',
                    'DISEASE GROUPING 3','DISEASE GROUPING 4','DISEASE GROUPING 5',
                    'DISEASE GROUPING 6', 'HTN','TEMPERATURE_MIN', 'OXYGEN_SATURATION_MIN',
                    'BLOODPRESSURE_DIASTOLIC_MAX', 'BLOODPRESSURE_SISTOLIC_MAX',
                    'HEART_RATE_MAX', 'RESPIRATORY_RATE_MAX', 'TEMPERATURE_MAX',
                    'OXYGEN_SATURATION_MAX', 'BLOODPRESSURE_DIASTOLIC_DIFF',
                    'BLOODPRESSURE_SISTOLIC_DIFF']

y_desafio3 = dados_limpos_desafio['ICU']
x_desafio3 = dados_limpos_desafio[x_columns].drop(colunas_a_dropar, axis = 1)

np.random.seed(73246)

x_desafio3_train, x_desafio3_test, y_desafio3_train, y_desafio3_test = train_test_split(x_desafio3, y_desafio3, stratify = y_desafio3)

modelo_desafio3_dummy = DummyClassifier(strategy = 'prior')
modelo_desafio3_dummy.fit(x_desafio3_train, y_desafio3_train)

y_desafio3_pred_dummy = modelo_desafio3_dummy.predict(x_desafio3_test)
accuracy_score(y_desafio3_test, y_desafio3_pred_dummy)

0.5340909090909091

In [ ]:
modelo_desafio3 = LogisticRegression(max_iter = 2000)
modelo_desafio3.fit(x_desafio3_train, y_desafio3_train)

y_desafio3_prediction = modelo_desafio3.predict(x_desafio3_test)
accuracy_score(y_desafio3_test, y_desafio3_prediction)

0.6704545454545454

Achei o resultado bem estranho... não mudou o *score* do **`DummyClassifier`**, como era esperado. Mas não entendo como pode melhorar o resultado do modelo de Regressão Logística. Para mim ter resultados melhores com menos informações foi inesperado.

Vamos aproveitar que o prazo para finalizar esse módulo é um pouco maior e prosseguir com algumas análises.

Quero verificar quais colunas ficaram com informações idênticas, ou seja, não vão influenciar em nada o resultado.

In [ ]:
lista_de_colunas = dados_limpos_desafio.columns
contador = 1
relat = pd.DataFrame(columns=['Num_Col','Coluna','Valor', 'Quant'])

for i in lista_de_colunas:
    values = dados_limpos_desafio[i].value_counts(dropna=False).keys().tolist()
    counts = dados_limpos_desafio[i].value_counts(dropna=False).tolist()
    coluna = i
    linha = {'Num_Col':contador,'Coluna':coluna, 'Valor':values, 'Quant':counts}
    relat = relat.append(linha, ignore_index=True)
    contador = contador + 1

mascara = relat.Quant.apply(lambda x: 352 in x)
relat_filtrado = relat[mascara]
len(relat_filtrado)

37

As 37 colunas que foram para o *DataFrame* **`relat_filtrado`** podem ser retiradas de nosso estudo, pois todas apresentaram os mesmos valores para todas as linhas, ou seja, não influenciam em nada o resultado.

In [ ]:
colunas_a_dropar = relat_filtrado['Coluna'].to_list()
colunas_a_dropar.pop(len(relat_filtrado)-1)

'WINDOW'

In [ ]:
y_desafio4 = dados_limpos_desafio['ICU']
x_desafio4 = dados_limpos_desafio[x_columns].drop(colunas_a_dropar, axis = 1).drop('ICU', axis = 1)


np.random.seed(73246)

x_desafio4_train, x_desafio4_test, y_desafio4_train, y_desafio4_test = train_test_split(x_desafio4, y_desafio4, stratify = y_desafio4)

modelo_desafio4_dummy = DummyClassifier(strategy = 'prior')
modelo_desafio4_dummy.fit(x_desafio4_train, y_desafio4_train)

y_desafio4_pred_dummy = modelo_desafio4_dummy.predict(x_desafio4_test)
accuracy_score(y_desafio4_test, y_desafio4_pred_dummy)

0.5340909090909091

In [ ]:
modelo_desafio4 = LogisticRegression(max_iter = 2000)
modelo_desafio4.fit(x_desafio4_train, y_desafio4_train)

y_desafio4_prediction = modelo_desafio4.predict(x_desafio4_test)
accuracy_score(y_desafio4_test, y_desafio4_prediction)

0.6590909090909091

#### **Desafio - Repasse a Aula**
---

>Essa aula não foi simples, repasse por ela e garanta que entendeu passo a passo

Como eu já havia visto que esse era um dos desafios, optei por assistir a aula pela primeira vez só prestando atenção, sem preencher o notebook como faço normalmente.

Na segunda vez eu fui acompanhando e preenchendo a parte de código do notebook com poucas anotações, exceto os desafios. Agora vou completar a parte de texto e resolver os dois desafios anteriores.

### **O que aprendemos nessa aula?**
---

>* A compreender o real problema do negócio;
>
>* A manipular os dados para resolver o problema real;
>
>* Como usar os modelos *Dummy* para comparar qualidade do modelo.

---
